oss_client.cpp

```diff
@@ -0,0 +1,907 @@
+
+bool oss_client::is_conn_setup_done(void)
+{
+    if (client::is_conn_setup_done())
+        return m_cluster_slots_state == slots_done;
+    else
+        return false;
+}
+
+bool oss_client::send_conn_setup_commands(struct timeval timestamp)
+{
+    bool sent = false;
+
+    if (m_config->authenticate && m_authentication != auth_done) {
+        if (m_authentication == auth_none) {
+            benchmark_debug_log("sending authentication command.\n");
+            MAIN_CONNECTION->protocol->authenticate(m_config->authenticate);
+            MAIN_CONNECTION->pipeline->push(new request(rt_auth, 0, &timestamp, 0));
+            m_authentication = auth_sent;
+            sent = true;
+        }
+    }
+    if (m_config->select_db && m_db_selection != select_done) {
+        if (m_db_selection == select_none) {
+            benchmark_debug_log("sending db selection command.\n");
+            MAIN_CONNECTION->protocol->select_db(m_config->select_db);
+            MAIN_CONNECTION->pipeline->push(new request(rt_select_db, 0, &timestamp, 0));
+            m_db_selection = select_sent;
+            sent = true;
+        }
+    }
+
+    if (m_cluster_slots_state == slots_none) {
+        benchmark_debug_log("sending cluster slots command.\n");
+        MAIN_CONNECTION->protocol->write_command_cluster_slots();
+        MAIN_CONNECTION->pipeline->push(new request(rt_cluster_slots, 0, &timestamp, 0));
+        m_cluster_slots_state = slots_sent;
+        sent = true;
+    }
+
+    return sent;
+}
+
+void oss_client::set_slots_to_shard() {
+    unsigned int num_of_conn = m_connections.size();
+
+    for (unsigned int i=0; i<num_of_conn; i++) {
+        shard_connection* conn = m_connections[i];
+
+        for (int j=conn->min_slots_range; j<=conn->max_slots_range; j++)
+            m_slots_to_shard[j] = i;
+    }
+}
+
+void oss_client::create_cluster_connection(protocol_response *r) {
+    // run over response and create connections
+    for (unsigned int i=0; i<r->get_mbulk_value()->mbulk_array.size(); i++) {
+        // create connection
+        mbulk_element* shard = r->get_mbulk_value()->mbulk_array[i];
+
+        int min_slot = strtol(shard->mbulk_array[0]->value, NULL, 10);
+        int max_slot = strtol(shard->mbulk_array[1]->value, NULL, 10);
+
+        // hostname/ip
+        mbulk_element* mbulk_addr_el = shard->mbulk_array[2]->mbulk_array[0];
+        char* addr = (char*) malloc(mbulk_addr_el->value_len + 1);
+        memcpy(addr, mbulk_addr_el->value, mbulk_addr_el->value_len);
+        addr[mbulk_addr_el->value_len] = '\0';
+
+        // port
+        mbulk_element* mbulk_port_el = shard->mbulk_array[2]->mbulk_array[1];
+        char* port = (char*) malloc(mbulk_port_el->value_len + 1);
+        memcpy(port, mbulk_port_el->value, mbulk_port_el->value_len);
+        port[mbulk_port_el->value_len] = '\0';
+
+        // if it is the main connection, no need to add connection, just update range
+        if (strcmp(addr, m_config->server) == 0) {
```

Comment:
There could be a few shards on the same server, you need to compare ports as well.